File: R/huanjie/stats.R
benjamini_hochberg <- function(pvalues, FDR=0.05){
    if(length(pvalues)==0)
        return(NA)
    real_pvalues <- pvalues[!(is.infinite(pvalues) | is.nan(pvalues) | is.na(pvalues))]
    sorted_values = sort(real_pvalues)
    critical_values = seq(1, length(sorted_values)) / length(sorted_values) * FDR
    idx = sorted_values < critical_values
    if(sum(idx)==0)
        return(NaN)
    else
        return(max(which(idx)))
}

File: python/huanjie/stats.py
import numpy as np


def benjamini_hochberg(pvalues, FDR=0.05):
    pvalues = np.array(pvalues)
    if pvalues.size == 0:
        return np.nan
    sorted_values = np.sort(pvalues[np.logical_not(np.logical_or(np.isnan(pvalues), np.isinf(pvalues)))], axis=None)
    critical_values = np.arange(1, len(sorted_values) + 1) / len(sorted_values) * FDR
    idx = np.argwhere(sorted_values < critical_values).flatten()
    if idx.size == 0:
        return np.nan
    else:
        return sorted_values[idx[-1]]
